File: context/ai/induction/sdk/bun-server/InductionSDK.ts
import { Profile, Action, Transformation } from "./models";
import { ProfileFetcher, Environment } from "./services";

export class InductionSDK {
  private readonly HEADER_NAME = "X-Induction-Profile-ID";

  constructor(private fetcher: ProfileFetcher) {}

  async instrumentRequest(url: string, init?: RequestInit): Promise<Response> {
    if (Environment.isProduction()) {
      return fetch(url, init);
    }

    const headers = new Headers(init?.headers);
    const profileId = headers.get(this.HEADER_NAME);

    if (profileId) {
      const profile = await this.fetcher.fetch(profileId);
      if (profile) {
        return this.applyProfile(url, init, profile);
      }
    }

    return fetch(url, init);
  }

  private async applyProfile(url: string, init: RequestInit | undefined, profile: Profile): Promise<Response> {
    switch (profile.action) {
      case Action.CallReal:
        return fetch(url, init);

      case Action.MockWhole:
        const body = profile.mockResponse || "";
        return new Response(body, {
          status: 200,
          statusText: "OK",
          headers: { "Content-Type": "application/json" },
        });

      case Action.Mutate:
        const response = await fetch(url, init);
        if (response.ok) {
          try {
            const data = await response.json();
            const mutatedData = this.applyMutations(data, profile.mutations || []);
            return new Response(JSON.stringify(mutatedData), {
              status: response.status,
              statusText: response.statusText,
              headers: response.headers,
            });
          } catch (e) {
            console.error(`[SDK] Cannot mutate body for profile ${profile.id}:`, e);
            return response;
          }
        }
        return response;

      case Action.Exception:
        throw new Error(profile.exceptionMessage || "Induced Exception");

      case Action.Delay:
        const delay = profile.delayMs || 1000;
        console.log(`[SDK] Inducing delay of ${delay}ms`);
        await new Promise((resolve) => setTimeout(resolve, delay));
        return fetch(url, init);

      default:
        return fetch(url, init);
    }
  }

  private applyMutations(data: any, mutations: Transformation[]): any {
    try {
      const result = { ...data };
      for (const m of mutations) {
        // Simple field replacement
        result[m.field] = m.value;
      }
      return result;
    } catch (e) {
      console.error(`[SDK] Failed to apply mutations:`, e);
      return data;
    }
  }
}

File: context/ai/induction/sdk/src/main/scala/com/induction/sdk/InductionSDK.scala
package com.induction.sdk

import sttp.client3.*
import upickle.default.*
import sttp.model.RequestMetadata
import sttp.model.Method
import sttp.model.Uri

case class SimpleRequestMetadata(
    method: Method,
    uri: Uri,
    headers: Seq[sttp.model.Header]
) extends RequestMetadata

class InductionSDK(fetcher: ProfileFetcher, backend: SttpBackend[Identity, Any] = HttpClientSyncBackend()):
  private val HEADER_NAME = "X-Induction-Profile-ID"

  def instrumentRequest[T](request: Request[T, Any]): Response[T] =
    if Environment.isProduction then
      request.send(backend)
    else
      val profileId = request.headers.find(_.name == HEADER_NAME).map(_.value)
      profileId match
        case Some(id) =>
          fetcher.fetch(id) match
            case Some(profile) => applyProfile(request, profile)
            case None => request.send(backend)
        case None =>
          request.send(backend)

  private def applyProfile[T](request: Request[T, Any], profile: Profile): Response[T] =
    profile.action match
      case Action.CallReal =>
        request.send(backend)

      case Action.MockWhole =>
        val body = profile.mockResponse.getOrElse("")
        createMockResponse(request, body)

      case Action.Mutate =>
        val response = request.send(backend)
        response.body match
          case Right(bodyStr: String) =>
            val mutatedBody = applyMutations(bodyStr, profile.mutations)
            response.copy(body = Right(mutatedBody).asInstanceOf[T])
          case _ =>
            println(s"[SDK] Cannot mutate non-string or error body for profile ${profile.id}")
            response

      case Action.Exception =>
        throw new RuntimeException(profile.exceptionMessage.getOrElse("Induced Exception"))

      case Action.Delay =>
        val delay = profile.delayMs.getOrElse(1000L)
        println(s"[SDK] Inducing delay of ${delay}ms")
        Thread.sleep(delay)
        request.send(backend)

  private def createMockResponse[T](request: Request[T, Any], body: String): Response[T] =
    Response(
      body = Right(body).asInstanceOf[T],
      code = sttp.model.StatusCode.Ok,
      statusText = "OK",
      headers = Nil,
      history = Nil,
      request = SimpleRequestMetadata(request.method, request.uri, request.headers)
    )

  private def applyMutations(json: String, mutations: List[Transformation]): String =
    try {
      var data = ujson.read(json)
      mutations.foreach { m =>
        // Simple field replacement. For nested fields, a more complex logic would be needed.
        data(m.field) = m.value
      }
      ujson.write(data)
    } catch {
      case e: Exception =>
        println(s"[SDK] Failed to apply mutations: ${e.getMessage}")
        json
    }
